File: solutions/1570-final-prices-with-a-special-discount-in-a-shop/solution.py
class Solution:
    def finalPrices(self, prices: List[int]) -> List[int]:
        # (price, index)
        s = []

        for i in range(len(prices)):
            price = prices[i]

            while s and price <= s[-1][0]:
                (prev_price, prev_index) = s.pop()
                prices[prev_index] = prev_price - price

            s.append((price, i))

        return prices

File: solutions/1570-final-prices-with-a-special-discount-in-a-shop/solution.js
/**
 * @param {number[]} prices
 * @return {number[]}
 */
var finalPrices = function(prices) {
    const stack = [];
    
    for(let i = 0; i < prices.length; i++){
        const price = prices[i];
        
        while(stack.length > 0 && price <= stack[stack.length - 1][0]){
            let old = stack.pop();
            
            prices[old[1]] = prices[old[1]] - price;
        }
        
        stack.push([price, i])    
    }
    
    return prices;
};
